### DISCUSSION TITLE
Question about lock_ttl and until_and_while_executing

### DISCUSSION THREAD

**Discussion by ghost** [mhenrixon/sidekiq-unique-jobs#731] [Q&A]
I inherited a job set up as:
```
 sidekiq_options retry: 3,
                  lock: :until_and_while_executing,
                  lock_timeout: 10.minutes,
                  lock_ttl: 20.minutes,
                  on_conflict: {
                    client: :log,
                    server: :reject
                  }
```

These jobs take a few seconds to run, and when a large number of them queue up the time spent on the queue can exceed the `lock_ttl`. When that happens, the jobs appear to no-op as soon as they are picked up for processing. I'm not seeing errors, just jobs instantly completing without doing any work.

The `lock_ttl` was added as part of an effort to address orphaned locks, so I think the solution here is straightforward -- remove the `lock_ttl` and let the reaper clean up any orphaned locks.

What I'm unclear on is what the intended behavior is for this use case -- while it makes sense to me that these jobs might not process after the `lock_ttl` expires, I might have expected something like an error.

Can anyone shed some insight on this?

**Comment by mhenrixon** [comment]
First, `lock_timeout: 10.minutes` might mean that one of your workers will be waiting for 10 minutes to push a job to Sidekiq.

To be honest, the timeout should likely only be used for the server process (when your perform action is called).

You want this number to be as low as possible. Try a second or two!

Hard to say what is going wrong, but in #730, it was discovered that `redis-namespace` is not working with keyword arguments, and there is an old PR from 2015 that potentially fixes the issue: https://github.com/resque/redis-namespace/pull/104.

**Reply by ghost** [reply]
Regarding `lock_timeout` -- I was 99% sure that was the case and would not have set that value myself, but I also wasn't aware of existing problems caused by it. I figured I'd post it as-is and see if it got the reaction I expected. It did and I'll fix it, thank you 😆 .

For the other issue -- we aren't using `redis-namespace`. I should have mentioned this in my original post but I noticed this behavior first on v7.1.23, and migrated to v7.1.26 yesterday to see if there was any difference. For both versions unique jobs process fine until they've been waiting to process longer than their `lock_ttl`.

I can try and debug this next week, I was just struggling with what the expected behavior would be when the `lock_ttl` expired and the job was still waiting to execute.

**Reply by ghost** [reply]
Also to clarify -- jobs are not going to the dead set as reported in #730, they are simply completing without executing their logic (as best I can tell).

**Reply by mhenrixon** [reply]
This is normal for this job. When the server tries to execute your job, it likely times out and the job is lost. 

If you want to retry the job you should have `on_conflict: { server: :raise }`which would let the sidekiq retry mechanism do it's job.

**Reply by egeek** [reply]
We're seeing the same behaviour - https://github.com/mperham/sidekiq/issues/5614

`sidekiq_options lock: :until_and_while_executing, on_conflict: :log`

`config.lock_ttl = 2.minutes`

When a job is queued, it has 2 minutes to run, or it will be dropped from the queue with a no-op. 

Test code:

```
  class SleepySchedulerJob < SidekiqJob
    include Sidekiq::Worker
    sidekiq_options queue: :default, retry: 0, backtrace: true
    sidekiq_options lock: :until_and_while_executing, on_conflict: :log

    def perform
      iterations = 200
      batch_size = 200

      (1..iterations).each do |batch|
        start_id = batch * batch_size
        end_id = start_id + batch_size
        ids = (start_id..end_id).map { |x| [x] }

        ::SleepyJob.perform_bulk(ids)
    end
  end
```

and 

```
  class SleepyJob < SidekiqJob
    sidekiq_options queue: :default, retry: 0, backtrace: true
    sidekiq_options lock: :until_and_while_executing, on_conflict: :log

    def perform(id)
      sleep 15
    end
  end
```

Maybe this is the expected behavour? Where a job is only allowed to stay in the queue until `lock_ttl` expires?

**Reply by pboling** [reply]
@mhenrixon We're also seeing jobs disappear when we have a short `lock_ttl`, and after reading this thread I think this is intended.  The documentation sort of hints at this, but it wasn't clear until I came here.

I had assumed that the expiry of the lock was the expiry of _only_ the lock, and the job would continue on, but without uniqueness protection.  This was an invalid assumption.

This leads me to another question...

With `lock: :until_and_while_executing` does it use a single lock to track both until and while?  What I'm asking is *subtle*... so let me give an example.

I'm interested in having a job be unique in the queue, and then as it moves to executing I want it to execute uniquely, _however_ while executing I want to be able to enqueue another identical job, _priming the pump_ if you will.  At any given time I want at most one in the queue, and one being executed.  My suite of uniqueness gems for resque work this way by having two separate key spaces, one for queue time and one for run time.  See [resque-unique_at_runtime](https://github.com/pboling/resque-unique_at_runtime), [resque-unique_in_queue](https://github.com/pboling/resque-unique_in_queue), and [resque-unique_by_arity](https://github.com/pboling/resque-unique_by_arity).

**Reply by pboling** [reply]
Found the answer here https://github.com/mhenrixon/sidekiq-unique-jobs/discussions/660

It is two different locks, one for queue-time and one for run-time.

**Reply by pboling** [reply]
Separately it does follow from the current documentation of `lock_ttl`, but it is easy to get distracted by the behavior at queue-time... 

The same `lock_ttl` value also *still* has a runtime behavior (i.e. it is sort-of overloaded in the scenario of `until_and_while_executing`), which is to kill the lock after the ttl expires.

So if set for `5.minutes`, and I have a job that takes more than 5 minutes to run, the (uniqueness) lock expires after 5 minutes, allowing an _identical_ job to begin executing.  Because in an `until_and_while_executing` configuration the point is to be able to have one executing and one queued at all times, if you use `lock_ttl` it means that your jobs must both:
1. begin executing (i.e. be removed from the queue) faster than the `lock_ttl` or they will be vanished (not even dead, just disappeared).
2. finish executing faster than the `lock_ttl` or they will allow the same job in the queue to start up alongside them.

It's the logical conclusion of expiring a lock with a TTL, but it is still hard to think about, and it might be worth mentioning that when using `lock: :until_and_while_executing` you almost certainly *do not want* a `lock_ttl`, unless you set it to `nil`.

**Reply by mhenrixon** [reply]
`lock_ttl` and `lock_timeout` should likely not be used on the first lock since I added the reaper. The reaper should make sure that no leftovers are kept in redis. 

For lock_ttl, it should kick in as execution starts, but I have been too chicken to break this for people, perhaps in the next major release.

**Reply by pboling** [reply]
> `lock_ttl` and `lock_timeout` should likely not be used on the first lock since I added the reaper

when you say "first lock" are you referring to the queue-time lock?

> For lock_ttl, it should kick in as execution starts

It seems to currently do this for the runtime lock (we're still on v7.1), so your phrasing is confusing.

In other words, I have a job with a lock_ttl of 5 minutes, and I had two identical jobs running, and they started 5 minutes apart.  There is always one in the queue to be picked up, so it clearly waited to pick it up until the lock_ttl had expired the run-time lock.

So what have you been too chicken to break? (I.e. what might change in the next major release?)

**Reply by ghost** [reply]
@pboling Thank you.

This has been nagging at me for ages, and I believe the distinction you drew between the two separate `lock_ttl`s was the missing piece for me.

**Reply by pboling** [reply]
@mhenrixon 

> For lock_ttl, it should kick in as execution starts, but I have been too chicken to break this for people, perhaps in the next major release.

Can you elaborate on the difference between what it "should do" and what it currently does, and what might break if you change the behavior?

I am particularly confused because "For lock_ttl, it should kick in as execution starts" this is the apparent behavior we see on v7.1.
